state/core/src/main/java/com/expediagroup/streamplatform/streamregistry/state/EntityViewUpdater.java

```diff
@@ -51,44 +54,51 @@
     }
   }
 
+  <K extends Entity.Key<S>, S extends Specification> Optional<Entity<K, S>> purge(K key) {
+    val oldEntity = (Entity<K, S>) entities.remove(deleted(key));
+    log.debug("Purged entity for {}", key);
+    return Optional.ofNullable(oldEntity);
+  }
+
   private <K extends Entity.Key<S>, S extends Specification> Entity<K, S> update(SpecificationEvent<K, S> event) {
-    val oldEntity = (Entity<K, S>) entities.get(event.getKey());
+    val oldEntity = (Entity<K, S>) entities.get(existing(event.getKey()));
     val status = Optional
         .ofNullable(oldEntity)
         .map(Entity::getStatus)
         .orElseGet(DefaultStatus::new);
     val entity = new Entity<>(event.getKey(), event.getSpecification(), status);
-    entities.put(event.getKey(), entity);
+    entities.put(existing(event.getKey()), entity);
     log.debug("Updated {} with {}", event.getKey(), event.getSpecification());
     return oldEntity;
   }
 
   private <K extends Entity.Key<S>, S extends Specification> Entity<K, S> update(StatusEvent<K, S> event) {
-    val oldEntity = (Entity<K, S>) entities.get(event.getKey());
+    val oldEntity = (Entity<K, S>) entities.get(existing(event.getKey()));
     if (oldEntity == null) {
       log.warn("Received status {} non existent entity {}", event.getStatusEntry().getName(), event.getKey());
       return null;
     }
     val entity = new Entity<>(event.getKey(), oldEntity.getSpecification(), oldEntity.getStatus().with(event.getStatusEntry()));
-    entities.put(event.getKey(), entity);
+    entities.put(existing(event.getKey()), entity);
     log.debug("Updated {} with {}", event.getKey(), event.getStatusEntry());
     return oldEntity;
   }
 
   private <K extends Entity.Key<S>, S extends Specification> Entity<K, S> delete(SpecificationDeletionEvent<K, S> event) {
-    val oldEntity = (Entity<K, S>) entities.remove(event.getKey());
+    val oldEntity = (Entity<K, S>) entities.remove(existing(event.getKey()));
```

Comment:
Yea, here - if the `deleted` flag was on `Entity` then you wouldn't need to remove/add, you'd just replace/mutate the existing entity with the deleted version.

Reply:
Coolios, makes sense.